Move log download logic into dataflash

peterbarker opened the pull request.

peterbarker commented on libraries/GCS_MAVLink/GCS.h:
```diff
@@ -129,6 +129,13 @@ class GCS_MAVLINK
     // call to reset the timeout window for entering the cli
     void reset_cli_timeout();
 
+    bool is_usb() { return chan == MAVLINK_COMM_0; }
```
So "is_high_bandwidth()" for the time being?  That assumption was there before.

peterbarker replied:
Of course, then we end up with this:

```
    if (link.is_high_bandwidth() && hal.gpio->usb_connected()) {
```

Which is even  worse than what was there :-)

Oh well.

OXINARF replied:
What about `is_console`? I don't like it very much either, but I think that is the only safe assumption we can make. Even high bandwidth isn't correct, because in a Linux all channels may be high bandwidth.

We shouldn't change now, but is there any reason not to base the check on the UART baudrate instead of anything else?

peterbarker replied:
In linux `is_high_bandwidth` can return true for all of them :-)

baudrate isn't the be-all-and-end-all of high-bandwidth (921,600 but 80% packet loss!) - but it would certainly suck less than "are we on COMM_0".

My original `is_usb` was just putting into code what was in comments - it wouldn't have been accurate on Linux, of course.

I suggest we leave in as "is_high_bandwidth" for the time being and come back and do better later on the implementation - probably based, as you say, on baud rate.

peterbarker commented on AntennaTracker/GCS_Mavlink.cpp:
```diff
@@ -376,6 +376,8 @@ GCS_MAVLINK_Tracker::data_stream_send(void)
         return;
     }
 
+    tracker.DataFlash.handle_log_send(*this);
```
Darn.  Will fix.

peterbarker commented:
Fixed the mixed-commits and the commit message.

This was originally going to be two separate PRs, the second based on the first.  Got a little carried away.  Again.

Pushed fixes up.

OXINARF commented:
Sub isn't removing the *in_log_download* variable.

I guess you don't want to squash this up, but can we at least remove *Copter: correct log restart after log erase* and *Sub: correct log restart after log erase*? The last one isn't doing anything, and the first one would also apply to Tracker, Rover and Plane, but there aren't commits for those... And by the way, this was changed by you in previous PR #5712, which looks correct to me - and is also the final behavior of this PR.

peterbarker commented:
On Tue, 20 Jun 2017, Francisco Ferreira wrote:

> Sub isn't removing the in_log_download variable.

Fixed.

> I guess you don't want to squash this up, but can we at least remove Copter:
> correct log restart after log erase and Sub: correct log restart after log

Squashed them up.  I was trying to avoid mixing "move code from here to 
there" with behaviour changes.  Since I failed on the other vehicle types, 
we may as well be consistently wrong.

OXINARF commented:
@peterbarker Have you seen my [comment](https://github.com/ArduPilot/ardupilot/pull/6456#discussion_r123005386)?

I think this is probably good to go, although I just noticed that only Copter and Sub care about *in_log_download* when starting logging.

peterbarker commented:
@OXINARF Consistency is something we're going to be able to get soon on all of this logging stuff.  Really looking forward to `should_log` NOT modifying system state by starting logging, because nothing with `should` in the name should do that, IMO :-)

OXINARF commented:
I agree with everything you wrote just now, so merged, thanks!
